From 9e0872cfbebd9630288d3708b08c825b6d8e3e1e Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:38:14 -0400
Subject: [alsa] pcm_dmaengine: Correct support for 24bits physical sample widths

Message-id: <140972823573-379-patchset-jkysela@redhat.com>
Patchwork-id: 91075
O-Subject: [RHEL 7.1 PATCH 378/442] ALSA: pcm_dmaengine: Correct support for 24bits physical sample widths
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

In case of _3LE/_3BE formats the samples are stored in 3 consecutive bytes
without padding it to 4 bytes. This means that the DMA needs to be able to
support 3 bytes word length in order to read/write the samples from memory
correctly. Originally the code treated 24 bits physical length samples as
they were 32 bits which leads to corruption when playing or recording audio.

The hw.formats field has already been prepared to exclude formats not
supported by the DMA engine in use, which means that only on platforms where
3 bytes is supported by the DMA will be able to use this format.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@ti.com>
Acked-by: Vinod Koul <vinod.koul@intel.com>
Acked-by: Takashi Iwai <tiwai@suse.de>
Signed-off-by: Mark Brown <broonie@linaro.org>

(picked from kernel: 75f850fc229e090d3073708f55ca628db650b64e)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/soc/soc-dmaengine-pcm.c b/sound/soc/soc-dmaengine-pcm.c
index d5611ec..6542c40 100644
--- a/sound/soc/soc-dmaengine-pcm.c
+++ b/sound/soc/soc-dmaengine-pcm.c
@@ -72,6 +72,8 @@ int snd_hwparams_to_dma_slave_config(const struct snd_pcm_substream *substream,
   buswidth = DMA_SLAVE_BUSWIDTH_1_BYTE;
  else if (bits == 16)
   buswidth = DMA_SLAVE_BUSWIDTH_2_BYTES;
+ else if (bits == 24)
+  buswidth = DMA_SLAVE_BUSWIDTH_3_BYTES;
  else if (bits <= 32)
   buswidth = DMA_SLAVE_BUSWIDTH_4_BYTES;
  else
-- 
1.7.1